### PR TITLE
Fixes #15233 - add index to `katello_repository_rpms`

### PR DESCRIPTION
**Pull request by beav**
The CVV page was calling
`/katello/api/v2/content_view_versions?content_view_id=<id>`, which includes,
among other things, a package count for all repos in each CVV.

This is found via the following query:

```sql
SELECT Count(*)
FROM   "katello_rpms"
       INNER JOIN "katello_repository_rpms"
               ON "katello_repository_rpms"."rpm_id" = "katello_rpms"."id"
WHERE  "katello_repository_rpms"."repository_id" IN (SELECT
       "katello_repositories"."id"
                                                     FROM
       "katello_repositories"
                                                     WHERE
              "katello_repositories"."content_view_version_id" = $1
              AND ( environment_id IS NULL ))
```

This query induced a sequential scan on `katello_repository_rpms` (see
https://explain.depesz.com/s/sMy).

There was an index for `repository_id, rpm_id` for the table, but not just
`repository_id`. This patch adds a migration to create the index.

Times per-query should be reduced from 300+ msec to 20-30 msec. The explain
plan with the index is available at https://explain.depesz.com/s/epd9.

_NOTE_: when testing, you will need to ensure `katello_repository_rpms` has
enough rows for the profiler to prefer the index over a seq scan. I used an
existing DB that was a shade under 4 million rows for the table. If the table
is too small, both explain plans (with and without the index) will look the
same.

### CODE REVIEW

**Comment by beav** [comment]
PR is purposefully >72 char due to SQL example.

@ehelms this should fix the issue you found when testing my other PR. You can do a simple `create index repo_idx on katello_repository_rpms (repository_id);` if you want to try it out by hand on an existing system (just remember to drop the index after testing so it doesn't conflict later on).

Many thanks to @jlsherrill for initial pointer with getting started on postgres explain plans.

**Comment by beav** [comment]
I added a second commit for some unrelated rubocop fixes that were making this PR fail jenkins (four whitespace alterations).

**Comment by jlsherrill** [comment]
Would you be open to going ahead and adding this for errata and possibly
the other content types?

PR is purposefully >72 char due to SQL example.

@ehelms <https://github.com/ehelms> this should fix the issue you found
when testing my other PR. You can do a simple create index repo_idx on
katello_repository_rpms (repository_id); if you want to try it out by hand
on an existing system (just remember to drop the index after testing so it
doesn't conflict later on).

Many thanks to @jlsherrill <https://github.com/jlsherrill> for initial
pointer with getting started on postgres explain plans.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Katello/katello/pull/6086#issuecomment-222556858>, or mute
the thread
<https://github.com/notifications/unsubscribe/AAYHRWYgCMjR4j2mXGc2MZTjG3iMqaTtks5qG1c4gaJpZM4IqCo8>
.

**Comment by beav** [comment]
@jlsherrill yeah that is probably a good idea. We can assume that the other tables will get larger over time and eventually cause similar issues.

Total list is

```
katello_repository_docker_manifests
katello_repository_errata
katello_repository_ostree_branches
katello_repository_package_groups
katello_repository_puppet_modules
katello_repository_rpms
```

I will add the additional indexes tomorrow.

**Comment by beav** [comment]
PR is updated per @jlsherrill comments

**Comment by ehelms** [comment]
This reduced load time to 2.29 seconds with 1.1 of it spent in active record. :+1:

File: db/migrate/20160530184400_add_repo_id_indexes.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class AddRepoIdIndexes < ActiveRecord::Migration
+  def up
+    add_index :katello_repository_docker_manifests, :repository_id
+    add_index :katello_repository_errata, :repository_id
+    add_index :katello_repository_ostree_branches, :repository_id
+    add_index :katello_repository_package_groups, :repository_id
+    add_index :katello_repository_puppet_modules, :repository_id
+    add_index :katello_repository_rpms, :repository_id
+  end
+end
```
Comment (komidore64) [inline db/migrate/20160530184400_add_repo_id_indexes.rb]: shouldn't we also include a `#down` method? i realize that people will rarely if ever migrate backwards, but i feel it's good practice to make sure migrations are happy in both directions.

Comment (ehelms) [inline db/migrate/20160530184400_add_repo_id_indexes.rb]: The `up` could be made into a `change` I think and cover both cases?

Comment (jlsherrill) [inline db/migrate/20160530184400_add_repo_id_indexes.rb]: +1 to change

**Comment by beav** [comment]
updated for `s/up/change/`

**Comment by komidore64** [comment]
:+1:

**Comment by beav** [comment]
@ehelms @jlsherrill is this ready for ACK? it contains a rubocop fix that I'd like to get into master, so other PRs can rebase and get green on jenkins.

**Comment by jlsherrill** [comment]
ACK

**Pull request merged**